Crashes out with an error when the cache is valid but the cache file is missing

If your CI server has purged your `.cache/HASH_ID.json` file (might be `.text` or `.buffer`, etc) but the `.cache/HASH_ID` is still there and valid (Vercel seems to be doing this on 11ty-website), Fetch will throw an error on a missing cache file. It would be better to refetch and repopulate the cache with new content when this happens.

```
[11ty] Original error stack trace: Error: Cannot find module '/vercel/path0/.cache/eleventy-fetch/eleventy-fetch-f3f8a31ff8038563d3f8944d6e7f1f.json'
[11ty] Require stack:
[11ty] - /vercel/path0/node_modules/@11ty/eleventy-fetch/src/AssetCache.js
[11ty] - /vercel/path0/node_modules/@11ty/eleventy-fetch/src/RemoteAssetCache.js
[11ty] - /vercel/path0/node_modules/@11ty/eleventy-fetch/eleventy-fetch.js
[11ty]     at Function._resolveFilename (node:internal/modules/cjs/loader:1225:15)
[11ty]     at Function._load (node:internal/modules/cjs/loader:1055:27)
[11ty]     at TracingChannel.traceSync (node:diagnostics_channel:322:14)
[11ty]     at wrapModuleLoad (node:internal/modules/cjs/loader:220:24)
[11ty]     at Module.require (node:internal/modules/cjs/loader:1311:12)
[11ty]     at require (node:internal/modules/helpers:136:16)
```

--- a/package.json
+++ b/package.json
@@ -48,6 +48,7 @@
 		"@11ty/eleventy-utils": "^2.0.1",
 		"@rgrove/parse-xml": "^4.2.0",
 		"debug": "^4.4.0",
+		"flatted": "^3.3.3",
 		"graceful-fs": "^4.2.11",
 		"p-queue": "6.6.2"
 	},

--- a/package.json
+++ b/package.json
@@ -45,9 +45,9 @@
 		"prettier": "^3.3.3"
 	},
 	"dependencies": {
+		"@11ty/eleventy-utils": "^2.0.1",
 		"@rgrove/parse-xml": "^4.2.0",
-		"debug": "^4.3.7",
-		"flat-cache": "^6.1.1",
+		"debug": "^4.4.0",
 		"graceful-fs": "^4.2.11",
 		"p-queue": "6.6.2"
 	},

--- a/package.json
+++ b/package.json
@@ -41,8 +41,8 @@
 	},
 	"homepage": "https://github.com/11ty/eleventy-fetch#readme",
 	"devDependencies": {
-		"ava": "^6.1.3",
-		"prettier": "^3.3.3"
+		"ava": "^6.2.0",
+		"prettier": "^3.5.3"
 	},
 	"dependencies": {
 		"@11ty/eleventy-utils": "^2.0.1",

--- a/src/AssetCache.js
+++ b/src/AssetCache.js
@@ -5,11 +5,9 @@ const { DateCompare } = require("@11ty/eleventy-utils");
 
 const FileCache = require("./FileCache.js");
 const Sources = require("./Sources.js");
-const DirectoryManager = require("./DirectoryManager.js");
 
 const debugUtil = require("debug");
 const debug = debugUtil("Eleventy:Fetch");
-const debugAssets = debugUtil("Eleventy:Assets");
 
 class AssetCache {
 	#source;
@@ -19,7 +17,6 @@ class AssetCache {
 	#cacheDirectory;
 	#cacheLocationDirty = false;
 	#directoryManager;
-	#rawContents = {};
 
 	constructor(source, cacheDirectory, options = {}) {
 		if(!Sources.isValidSource(source)) {
@@ -227,12 +224,15 @@ class AssetCache {
 	}
 
 	isCacheValid(duration = this.duration) {
-		if (!this.cachedObject || !this.cachedObject.cachedAt) {
-			// not cached
+		if(!this.cachedObject || !this.cachedObject?.cachedAt) {
 			return false;
 		}
 
-		return DateCompare.isTimestampWithinDuration(this.cachedObject.cachedAt, duration);
+		if(DateCompare.isTimestampWithinDuration(this.cachedObject?.cachedAt, duration)) {
+			return this.cache.hasContents(); // check file system to make files haven’t been purged.
+		}
+
+		return false;
 	}
 
 	get cachedObject() {

--- a/src/AssetCache.js
+++ b/src/AssetCache.js
@@ -1,9 +1,10 @@
 const fs = require("graceful-fs");
 const path = require("path");
-const { create: FlatCacheCreate } = require("flat-cache");
 const { createHash } = require("crypto");
 const debugUtil = require("debug");
+const { DateCompare } = require("@11ty/eleventy-utils");
 
+const FileCache = require("./FileCache.js");
 const Sources = require("./Sources.js");
 const DirectoryManager = require("./DirectoryManager.js");
 
@@ -18,7 +19,7 @@ class AssetCache {
 	#cacheDirectory;
 	#cacheLocationDirty = false;
 	#directoryManager;
-	#rawContents = {}
+	#rawContents = {};
 
 	constructor(source, cacheDirectory, options = {}) {
 		if(!Sources.isValidSource(source)) {
@@ -173,10 +174,11 @@ class AssetCache {
 
 	get cache() {
 		if (!this.#cache || this.#cacheLocationDirty) {
-			let cache = FlatCacheCreate({
-				cacheId: this.cacheFilename,
-				cacheDir: this.rootDir,
+			let cache = new FileCache(this.cacheFilename, {
+				dir: this.rootDir
 			});
+			cache.setDryRun(this.options.dryRun);
+			cache.setDirectoryManager(this.#directoryManager);
 
 			this.#cache = cache;
 			this.#cacheLocationDirty = false;
@@ -185,24 +187,7 @@ class AssetCache {
 	}
 
 	getDurationMs(duration = "0s") {
-		let durationUnits = duration.slice(-1);
-		let durationMultiplier;
-		if (durationUnits === "s") {
-			durationMultiplier = 1;
-		} else if (durationUnits === "m") {
-			durationMultiplier = 60;
-		} else if (durationUnits === "h") {
-			durationMultiplier = 60 * 60;
-		} else if (durationUnits === "d") {
-			durationMultiplier = 60 * 60 * 24;
-		} else if (durationUnits === "w") {
-			durationMultiplier = 60 * 60 * 24 * 7;
-		} else if (durationUnits === "y") {
-			durationMultiplier = 60 * 60 * 24 * 365;
-		}
-
-		let durationValue = parseInt(duration.slice(0, duration.length - 1), 10);
-		return durationValue * durationMultiplier * 1000;
+		return DateCompare.getDurationMs(duration);
 	}
 
 	getCachedContentsPath(type = "buffer") {
@@ -263,8 +248,6 @@ class AssetCache {
 		// the contents must exist before the cache metadata are saved below
 		fs.writeFileSync(contentPath, contents);
 		debug(`Writing ${contentPath}`);
-
-		this.cache.save();
 	}
 
 	async #getCachedContents(type) {
@@ -335,27 +318,7 @@ class AssetCache {
 			return false;
 		}
 
-		// in the cache and no duration
-		if (!duration || duration === "*") {
-			// no duration specified (plugin default is 1d, but if this is falsy assume infinite)
-			// "*" is infinite duration
-			return true;
-		}
-
-		debug("Cache check for: %o %o (duration: %o)", this.hash, this.source, duration);
-		debug("Cache object: %o", this.cachedObject);
-
-		let compareDuration = this.getDurationMs(duration);
-		let expiration = this.cachedObject.cachedAt + compareDuration;
-		let expirationRelative = Math.abs(Date.now() - expiration);
-
-		if (expiration > Date.now()) {
-			debug("Cache okay, expires in %o s (%o)", expirationRelative / 1000, new Date(expiration));
-			return true;
-		}
-
-		debug("Cache expired %o s ago (%o)", expirationRelative / 1000, new Date(expiration));
-		return false;
+		return DateCompare.isTimestampWithinDuration(this.cachedObject.cachedAt, duration);
 	}
 
 	get cachedObject() {

--- a/src/AssetCache.js
+++ b/src/AssetCache.js
@@ -1,13 +1,13 @@
 const fs = require("graceful-fs");
 const path = require("path");
 const { createHash } = require("crypto");
-const debugUtil = require("debug");
 const { DateCompare } = require("@11ty/eleventy-utils");
 
 const FileCache = require("./FileCache.js");
 const Sources = require("./Sources.js");
 const DirectoryManager = require("./DirectoryManager.js");
 
+const debugUtil = require("debug");
 const debug = debugUtil("Eleventy:Fetch");
 const debugAssets = debugUtil("Eleventy:Assets");
 
@@ -175,7 +175,8 @@ class AssetCache {
 	get cache() {
 		if (!this.#cache || this.#cacheLocationDirty) {
 			let cache = new FileCache(this.cacheFilename, {
-				dir: this.rootDir
+				dir: this.rootDir,
+				source: this.source,
 			});
 			cache.setDryRun(this.options.dryRun);
 			cache.setDirectoryManager(this.#directoryManager);
@@ -188,16 +189,6 @@ class AssetCache {
 
 	getDurationMs(duration = "0s") {
 		return DateCompare.getDurationMs(duration);
-	}
-
-	getCachedContentsPath(type = "buffer") {
-		if(type === "xml") {
-			type = "text";
-		} else if(type === "parsed-xml") {
-			type = "json";
-		}
-
-		return `${this.cachePath}.${type}`;
 	}
 
 	setDirectoryManager(manager) {
@@ -222,59 +213,14 @@ class AssetCache {
 			throw new Error("save(contents) expects contents (was falsy)");
 		}
 
-		this.cache.set(this.hash, {
-			cachedAt: Date.now(),
-			type: type,
-			metadata,
-		});
+		this.cache.set(type, contents, metadata);
 
-		let contentPath = this.getCachedContentsPath(type);
-
-		if (type === "json" || type === "parsed-xml") {
-			contents = JSON.stringify(contents);
-		}
-
-		this.#rawContents[type] = contents;
-
-		if(this.options.dryRun) {
-			debug(`Dry run writing ${contentPath}`);
-			return;
-		}
-
-		this.ensureDir();
-
-		debugAssets("[11ty/eleventy-fetch] Writing %o from %o", contentPath, this.source);
-
-		// the contents must exist before the cache metadata are saved below
-		fs.writeFileSync(contentPath, contents);
-		debug(`Writing ${contentPath}`);
+		// Dry-run handled downstream
+		this.#cache.save();
 	}
 
-	async #getCachedContents(type) {
-		let contentPath = this.getCachedContentsPath(type);
-
-		debug(`Fetching from cache ${contentPath}`);
-
-		if(this.source) {
-			debugAssets("[11ty/eleventy-fetch] Reading via %o", this.source);
-		} else {
-			debugAssets("[11ty/eleventy-fetch] Reading %o", contentPath);
-		}
-
-		if (type === "json" || type === "parsed-xml") {
-			return require(contentPath);
-		}
-
-		return fs.readFileSync(contentPath, type !== "buffer" ? "utf8" : null);
-	}
-
-	async getCachedContents(type) {
-		if(!this.#rawContents[type]) {
-			this.#rawContents[type] = this.#getCachedContents(type);
-		}
-
-		// already saved on this instance in-memory
-		return this.#rawContents[type];
+	getCachedContents() {
+		return this.#cache.getContents();
 	}
 
 	_backwardsCompatibilityGetCachedValue(type) {
@@ -288,10 +234,11 @@ class AssetCache {
 		return Buffer.from(this.cachedObject.contents);
 	}
 
-	async getCachedValue() {
+	getCachedValue() {
 		let type = this.cachedObject.type;
 
 		// backwards compat with old caches
+		// v6+ caches use `data` as internal property for contents
 		if (this.cachedObject.contents) {
 			return this._backwardsCompatibilityGetCachedValue(type);
 		}
@@ -299,12 +246,11 @@ class AssetCache {
 		if(this.options.returnType === "response") {
 			return {
 				...this.cachedObject.metadata?.response,
-				body: await this.getCachedContents(type),
+				body: this.getCachedContents(type),
 				cache: "hit",
 			}
 		}
 
-		// promise
 		return this.getCachedContents(type);
 	}
 
@@ -313,7 +259,7 @@ class AssetCache {
 	}
 
 	isCacheValid(duration = this.duration) {
-		if (!this.cachedObject) {
+		if (!this.cachedObject || !this.cachedObject.cachedAt) {
 			// not cached
 			return false;
 		}
@@ -322,7 +268,7 @@ class AssetCache {
 	}
 
 	get cachedObject() {
-		return this.cache.get(this.hash);
+		return this.cache.get();
 	}
 
 	// Deprecated
@@ -345,17 +291,18 @@ class AssetCache {
 	}
 
 	// for testing
-	hasCacheFiles() {
-		return fs.existsSync(this.cachePath) || fs.existsSync(this.getCachedContentsPath());
+	hasAnyCacheFiles() {
+		for(let p of this.#cache.getFilePaths()) {
+			if(fs.existsSync(p)) {
+				return true;
+			}
+		}
+		return false;
 	}
 
 	// for testing
 	async destroy() {
-		let paths = [];
-		paths.push(this.cachePath);
-		paths.push(this.getCachedContentsPath("json"));
-		paths.push(this.getCachedContentsPath("text"));
-		paths.push(this.getCachedContentsPath("buffer"));
+		let paths = this.#cache.getFilePaths();
 
 		await Promise.all(paths.map(path => {
 			if (fs.existsSync(path)) {

--- a/src/AssetCache.js
+++ b/src/AssetCache.js
@@ -195,19 +195,6 @@ class AssetCache {
 		this.#directoryManager = manager;
 	}
 
-	ensureDir() {
-		if (this.options.dryRun) {
-			return;
-		}
-
-		if(!this.#directoryManager) {
-			// standalone fallback (for tests)
-			this.#directoryManager = new DirectoryManager();
-		}
-
-		this.#directoryManager.create(this.cacheDirectory);
-	}
-
 	async save(contents, type = "buffer", metadata = {}) {
 		if(!contents) {
 			throw new Error("save(contents) expects contents (was falsy)");
@@ -216,42 +203,23 @@ class AssetCache {
 		this.cache.set(type, contents, metadata);
 
 		// Dry-run handled downstream
-		this.#cache.save();
+		this.cache.save();
 	}
 
 	getCachedContents() {
-		return this.#cache.getContents();
-	}
-
-	_backwardsCompatibilityGetCachedValue(type) {
-		if (type === "json") {
-			return this.cachedObject.contents;
-		} else if (type === "text") {
-			return this.cachedObject.contents.toString();
-		}
-
-		// buffer
-		return Buffer.from(this.cachedObject.contents);
+		return this.cache.getContents();
 	}
 
 	getCachedValue() {
-		let type = this.cachedObject.type;
-
-		// backwards compat with old caches
-		// v6+ caches use `data` as internal property for contents
-		if (this.cachedObject.contents) {
-			return this._backwardsCompatibilityGetCachedValue(type);
-		}
-
 		if(this.options.returnType === "response") {
 			return {
 				...this.cachedObject.metadata?.response,
-				body: this.getCachedContents(type),
+				body: this.getCachedContents(),
 				cache: "hit",
 			}
 		}
 
-		return this.getCachedContents(type);
+		return this.getCachedContents();
 	}
 
 	getCachedTimestamp() {
@@ -292,7 +260,7 @@ class AssetCache {
 
 	// for testing
 	hasAnyCacheFiles() {
-		for(let p of this.#cache.getFilePaths()) {
+		for(let p of this.cache.getFilePaths()) {
 			if(fs.existsSync(p)) {
 				return true;
 			}
@@ -302,7 +270,7 @@ class AssetCache {
 
 	// for testing
 	async destroy() {
-		let paths = this.#cache.getFilePaths();
+		let paths = this.cache.getFilePaths();
 
 		await Promise.all(paths.map(path => {
 			if (fs.existsSync(path)) {

--- a/src/DirectoryManager.js
+++ b/src/DirectoryManager.js
@@ -1,5 +1,5 @@
 const fs = require("node:fs");
-const debug = require("debug")("Eleventy:Assets");
+const debugAssets = require("debug")("Eleventy:Assets");
 
 class DirectoryManager {
   #dirs = new Set();
@@ -14,7 +14,7 @@ class DirectoryManager {
     }
 
     this.#dirs.add(dir);
-    debug("Creating directory %o", dir);
+    debugAssets("Creating directory %o", dir);
     fs.mkdirSync(dir, { recursive: true });
   }
 }

--- a/src/DirectoryManager.js
+++ b/src/DirectoryManager.js
@@ -1,7 +1,5 @@
 const fs = require("node:fs");
-const path = require("node:path");
-const debugUtil = require("debug");
-const debugAssets = debugUtil("Eleventy:Assets");
+const debug = require("debug")("Eleventy:Assets");
 
 class DirectoryManager {
   #dirs = new Set();
@@ -16,7 +14,7 @@ class DirectoryManager {
     }
 
     this.#dirs.add(dir);
-    debugAssets("[11ty/eleventy-fetch] Creating directory %o", dir);
+    debug("Creating directory %o", dir);
     fs.mkdirSync(dir, { recursive: true });
   }
 }

--- a/src/ExistsCache.js
+++ b/src/ExistsCache.js
@@ -1,0 +1,24 @@
+const fs = require("node:fs");
+// const debug = require("debug")("Eleventy:Assets");
+
+class ExistsCache {
+	#checks = new Map();
+	#count = 0;
+
+	set(target, value) {
+		this.#checks.set(target, Boolean(value));
+	}
+
+	exists(target) {
+		if(this.#checks.has(target)) {
+			return this.#checks.get(target);
+		}
+
+		let exists = fs.existsSync(target);
+		this.#count++;
+		this.#checks.set(target, exists);
+		return exists;
+	}
+}
+
+module.exports = ExistsCache;

--- a/src/FileCache.js
+++ b/src/FileCache.js
@@ -1,0 +1,83 @@
+const fs = require("node:fs");
+const path = require("node:path");
+
+const DirectoryManager = require("./DirectoryManager.js");
+const ExistsCache = require("./ExistsCache.js");
+
+let existsCache = new ExistsCache();
+
+class FileCache {
+	#directoryManager;
+	#data = {};
+	#dryRun = false;
+	#cacheDirectory = ".cache";
+	#counts = {
+		read: 0,
+		write: 0,
+	}
+
+	constructor(cacheId, options = {}) {
+		this.cacheId = cacheId;
+		if(options.dir) {
+			this.#cacheDirectory = options.dir;
+		}
+	}
+
+	setDryRun(val) {
+		this.#dryRun = Boolean(val);
+	}
+
+	setDirectoryManager(manager) {
+		this.#directoryManager = manager;
+	}
+
+	ensureDir() {
+		if (this.#dryRun || existsCache.exists(this.#cacheDirectory)) {
+			return;
+		}
+
+		if(!this.#directoryManager) {
+			// standalone fallback (for tests)
+			this.#directoryManager = new DirectoryManager();
+		}
+
+		this.#directoryManager.create(this.#cacheDirectory);
+	}
+
+	set(id, obj) {
+		if(this.#data[id] !== obj) {
+			this.#data[id] = obj;
+			this.save();
+		}
+	}
+
+	get fsPath() {
+		return path.join(this.#cacheDirectory, this.cacheId);
+	}
+
+	get(id) {
+		if(this.#data[id]) {
+			return this.#data[id];
+		}
+
+		if(!existsCache.exists(this.fsPath)) {
+			return;
+		}
+
+		this.#counts.read++;
+
+		let data = fs.readFileSync(this.fsPath, "utf8");
+		let json = JSON.parse(data);
+		this.#data[id] = json;
+		return json;
+	}
+
+	save() {
+		this.ensureDir(); // doesn’t add to counts (yet?)
+
+		this.#counts.write++;
+		fs.writeFileSync(this.fsPath, JSON.stringify(this.#data), "utf8");
+	}
+}
+
+module.exports = FileCache;

--- a/src/FileCache.js
+++ b/src/FileCache.js
@@ -189,7 +189,7 @@ class FileCache {
 			this.#counts.write++;
 
 			// the contents must exist before the cache metadata are saved below
-			fs.writeFileSync(this.contentsPath, contents);
+			fs.writeFileSync(this.contentsPath, this.#contents);
 			debug(`Writing ${this.contentsPath}`);
 		}
 

--- a/src/FileCache.js
+++ b/src/FileCache.js
@@ -131,6 +131,18 @@ class FileCache {
 		return Buffer.from(rawData.contents);
 	}
 
+	hasContents() {
+		if(this.#contents) {
+			return true;
+		}
+		if(!this.isSideLoaded() && this.get()?.data) {
+			return true;
+		} else if(this.get()?.contents) {
+			return true;
+		}
+		return existsCache.exists(this.contentsPath);
+	}
+
 	getContents() {
 		if(this.#contents) {
 			return this.#contents;

--- a/src/FileCache.js
+++ b/src/FileCache.js
@@ -1,5 +1,8 @@
 const fs = require("node:fs");
 const path = require("node:path");
+const debugUtil = require("debug");
+const debug = debugUtil("Eleventy:Fetch");
+const debugAssets = debugUtil("Eleventy:Assets");
 
 const DirectoryManager = require("./DirectoryManager.js");
 const ExistsCache = require("./ExistsCache.js");
@@ -7,19 +10,25 @@ const ExistsCache = require("./ExistsCache.js");
 let existsCache = new ExistsCache();
 
 class FileCache {
+	#source;
 	#directoryManager;
-	#data = {};
+	#metadata;
+	#contents;
 	#dryRun = false;
 	#cacheDirectory = ".cache";
+	#savePending = false;
 	#counts = {
 		read: 0,
 		write: 0,
-	}
+	};
 
-	constructor(cacheId, options = {}) {
-		this.cacheId = cacheId;
+	constructor(cacheFilename, options = {}) {
+		this.cacheFilename = cacheFilename;
 		if(options.dir) {
 			this.#cacheDirectory = options.dir;
+		}
+		if(options.source) {
+			this.#source = options.source;
 		}
 	}
 
@@ -44,39 +53,118 @@ class FileCache {
 		this.#directoryManager.create(this.#cacheDirectory);
 	}
 
-	set(id, obj) {
-		if(this.#data[id] !== obj) {
-			this.#data[id] = obj;
-			this.save();
-		}
+	isSideLoaded() {
+		return this.#metadata?.type === "buffer";
+	}
+
+	set(type, contents, extraMetadata = {}) {
+		this.#savePending = true;
+
+		this.#metadata = {
+			cachedAt: Date.now(),
+			type,
+			// source: this.#source,
+			metadata: extraMetadata,
+		};
+
+		this.#contents = contents;
 	}
 
 	get fsPath() {
-		return path.join(this.#cacheDirectory, this.cacheId);
+		return path.join(this.#cacheDirectory, this.cacheFilename);
 	}
 
-	get(id) {
-		if(this.#data[id]) {
-			return this.#data[id];
+	// only when side loaded (buffer content)
+	get contentsPath() {
+		return `${this.fsPath}.buffer`;
+	}
+
+	get() {
+		if(this.#metadata) {
+			return this.#metadata;
 		}
 
 		if(!existsCache.exists(this.fsPath)) {
 			return;
 		}
 
-		this.#counts.read++;
+		debug(`Fetching from cache ${this.contentsPath}`);
+		if(this.#source) {
+			debugAssets("[11ty/eleventy-fetch] Reading via %o", this.#source);
+		} else {
+			debugAssets("[11ty/eleventy-fetch] Reading %o", this.contentsPath);
+		}
 
+		this.#counts.read++;
 		let data = fs.readFileSync(this.fsPath, "utf8");
 		let json = JSON.parse(data);
-		this.#data[id] = json;
+		this.#metadata = json;
+
+		if(json.data) { // not side-loaded
+			this.#contents = json.data;
+		}
+
 		return json;
 	}
 
+	getContents() {
+		if(this.#contents) {
+			return this.#contents;
+		}
+		// Side loaded contents are embedded inside, but we check (for backwards compat)
+		if(!this.isSideLoaded() && this.get()?.data) {
+			return this.get()?.data;
+		}
+
+		if(!existsCache.exists(this.contentsPath)) {
+			return;
+		}
+
+		debug(`Fetching from cache ${this.contentsPath}`);
+		if(this.#source) {
+			debugAssets("[11ty/eleventy-fetch] Reading (side loaded) via %o", this.#source);
+		} else {
+			debugAssets("[11ty/eleventy-fetch] Reading (side loaded) %o", this.contentsPath);
+		}
+
+		this.#counts.read++;
+		let data = fs.readFileSync(this.contentsPath, null);
+		this.#contents = data;
+		return data;
+	}
+
 	save() {
+		if(this.#dryRun || !this.#savePending || this.#metadata && Object.keys(this.#metadata) === 0) {
+			return;
+		}
+
 		this.ensureDir(); // doesn’t add to counts (yet?)
 
+		// contents before metadata
+		if(this.isSideLoaded()) {
+			debugAssets("[11ty/eleventy-fetch] Writing %o (side loaded) from %o", this.contentsPath, this.#source);
+
+			this.#counts.write++;
+
+			// the contents must exist before the cache metadata are saved below
+			fs.writeFileSync(this.contentsPath, contents);
+			debug(`Writing ${this.contentsPath}`);
+		}
+
 		this.#counts.write++;
-		fs.writeFileSync(this.fsPath, JSON.stringify(this.#data), "utf8");
+		debugAssets("[11ty/eleventy-fetch] Writing %o from %o", this.fsPath, this.#source);
+		fs.writeFileSync(this.fsPath, JSON.stringify(Object.assign({}, this.#metadata, { data: this.#contents })), "utf8");
+		debug(`Writing ${this.fsPath}`);
+	}
+
+	// for testing
+	getFilePaths() {
+		let paths = new Set();
+		paths.add(this.fsPath);
+		if(this.isSideLoaded()) {
+			paths.add(this.contentsPath);
+		}
+		return Array.from(paths);
 	}
 }
 

--- a/src/FileCache.js
+++ b/src/FileCache.js
@@ -1,6 +1,8 @@
 const fs = require("node:fs");
 const path = require("node:path");
 const debugUtil = require("debug");
+const { parse } = require("flatted");
+
 const debug = debugUtil("Eleventy:Fetch");
 const debugAssets = debugUtil("Eleventy:Assets");
 
@@ -97,7 +99,18 @@ class FileCache {
 
 		this.#counts.read++;
 		let data = fs.readFileSync(this.fsPath, "utf8");
-		let json = JSON.parse(data);
+
+		let json;
+		// Backwards compatibility with previous caches usingn flat-cache and `flatted`
+		if(data.startsWith(`[["1"],`)) {
+			let flattedParsed = parse(data);
+			if(flattedParsed?.[0]?.value) {
+				json = flattedParsed?.[0]?.value
+			}
+		} else {
+			json = JSON.parse(data);
+		}
+
 		this.#metadata = json;
 
 		if(json.data) { // not side-loaded
@@ -107,13 +120,30 @@ class FileCache {
 		return json;
 	}
 
+	_backwardsCompatGetContents(rawData, type) {
+		if (type === "json") {
+			return rawData.contents;
+		} else if (type === "text") {
+			return rawData.contents.toString();
+		}
+
+		// buffer
+		return Buffer.from(rawData.contents);
+	}
+
 	getContents() {
 		if(this.#contents) {
 			return this.#contents;
 		}
+
 		// Side loaded contents are embedded inside, but we check (for backwards compat)
 		if(!this.isSideLoaded() && this.get()?.data) {
 			return this.get()?.data;
+		} else if(this.get()?.contents) {
+			// backwards compat with old caches
+			let normalizedContent = this._backwardsCompatGetContents(this.get(), this.#metadata.type);
+			this.#contents = normalizedContent; // set cache
+			return normalizedContent;
 		}
 
 		if(!existsCache.exists(this.contentsPath)) {

--- a/test/AssetCacheTest.js
+++ b/test/AssetCacheTest.js
@@ -44,15 +44,12 @@ test("AWS Lambda root directory resolves correctly", (t) => {
 test("Test a save", async (t) => {
 	let asset = new AssetCache("zachleat_twitter_followers", ".customcache");
 	let cachePath = normalizePath(asset.cachePath);
-	let jsonCachePath = normalizePath(asset.getCachedContentsPath("json"));
 
 	await asset.save({ followers: 10 }, "json");
 
-	t.truthy(fs.existsSync(jsonCachePath));
 	t.truthy(fs.existsSync(cachePath));
 
 	fs.unlinkSync(cachePath);
-	fs.unlinkSync(jsonCachePath);
 });
 
 test("Cache path should handle slashes without creating directories, issue #14", (t) => {
@@ -66,51 +63,35 @@ test("Uses `requestId` property when caching a promise", async (t) => {
 	let asset = new AssetCache(Promise.resolve(), ".customcache", {
 		requestId: "mock-display-url-2",
 	});
-	let cachePath = normalizePath(asset.cachePath);
-	let jsonCachePath = normalizePath(asset.getCachedContentsPath("json"));
 
 	await asset.save({ name: "Sophia Smith" }, "json");
 
-	t.truthy(fs.existsSync(jsonCachePath));
-
 	await asset.destroy();
 
-	t.falsy(fs.existsSync(cachePath));
-	t.falsy(fs.existsSync(jsonCachePath));
+	t.falsy(asset.hasAnyCacheFiles());
 });
 
 test("Uses `requestId` property when caching a function", async (t) => {
 	let asset = new AssetCache(function() {}, ".cache", {
 		requestId: "mock-function",
 	});
-	let cachePath = normalizePath(asset.cachePath);
-	let jsonCachePath = normalizePath(asset.getCachedContentsPath("json"));
-
 	await asset.save({ name: "Sophia Smith" }, "json");
-
-	t.truthy(fs.existsSync(jsonCachePath));
 
 	await asset.destroy();
 
-	t.falsy(fs.existsSync(cachePath));
-	t.falsy(fs.existsSync(jsonCachePath));
+	t.falsy(asset.hasAnyCacheFiles());
 });
 
 test("Uses `requestId` property when caching an async function", async (t) => {
 	let asset = new AssetCache(async function() {}, ".cache", {
 		requestId: "mock-async-function",
 	});
-	let cachePath = normalizePath(asset.cachePath);
-	let jsonCachePath = normalizePath(asset.getCachedContentsPath("json"));
 
 	await asset.save({ name: "Sophia Smith" }, "json");
 
-	t.truthy(fs.existsSync(jsonCachePath));
-
 	await asset.destroy();
 
-	t.falsy(fs.existsSync(cachePath));
-	t.falsy(fs.existsSync(jsonCachePath));
+	t.falsy(asset.hasAnyCacheFiles());
 });
 
 test("Uses filenameFormat", async (t) => {
@@ -122,18 +103,14 @@ test("Uses filenameFormat", async (t) => {
 	});
 
 	let cachePath = normalizePath(asset.cachePath);
-	let jsonCachePath = normalizePath(asset.getCachedContentsPath("json"));
 
 	t.truthy(cachePath.endsWith("/.cache/testing"));
-	t.truthy(jsonCachePath.endsWith("/.cache/testing.json"));
 
 	await asset.save({ name: "Sophia Smith" }, "json");
 
 	t.truthy(fs.existsSync(cachePath));
-	t.truthy(fs.existsSync(jsonCachePath));
 
 	await asset.destroy();
 
-	t.falsy(fs.existsSync(cachePath));
-	t.falsy(fs.existsSync(jsonCachePath));
+	t.falsy(asset.hasAnyCacheFiles());
 });

--- a/test/AssetCacheTest.js
+++ b/test/AssetCacheTest.js
@@ -114,3 +114,20 @@ test("Uses filenameFormat", async (t) => {
 
 	t.falsy(asset.hasAnyCacheFiles());
 });
+
+test("v5 flatted cache file", async (t) => {
+	let asset = new AssetCache("some-thing", "test", {
+		filenameFormat() {
+			// don’t include the file extension
+			return "v5flattedcachefile";
+		},
+	});
+
+	let cachePath = normalizePath(asset.cachePath);
+	t.truthy(cachePath.endsWith("test/v5flattedcachefile"));
+	t.truthy(asset.cachedObject.cachedAt);
+	t.truthy(asset.cachedObject.metadata);
+	t.is(asset.cachedObject.type, "json");
+	t.deepEqual(asset.getCachedValue(), undefined);
+	t.deepEqual(asset.getCachedContents(), undefined);
+});

--- a/test/QueueTest.js
+++ b/test/QueueTest.js
@@ -50,7 +50,7 @@ test("Double Fetch (dry run)", async (t) => {
 		dryRun: true,
 	});
 	// file is now accessible
-	t.false(forTestOnly.hasCacheFiles());
+	t.false(forTestOnly.hasAnyCacheFiles());
 });
 
 test("Double Fetch async function (dry run)", async (t) => {

--- a/test/RemoteAssetCacheTest.js
+++ b/test/RemoteAssetCacheTest.js
@@ -129,7 +129,7 @@ test("Fetching (dry run)!", async (t) => {
 	let buffer = await ac.fetch();
 	t.is(Buffer.isBuffer(buffer), true);
 	t.is(ac.fetchCount, 1);
-	t.false(ac.hasCacheFiles());
+	t.false(ac.hasAnyCacheFiles());
 });
 
 test("Fetching pass in URL", async (t) => {
@@ -237,9 +237,7 @@ test("Error with `cause`", async (t) => {
 		);
 		t.truthy(e.cause);
 	} finally {
-		try {
-			await asset.destroy();
-		} catch (e) {}
+		await asset.destroy();
 	}
 });
 

--- a/test/v5flattedcachefile
+++ b/test/v5flattedcachefile
@@ -1,0 +1,1 @@
+[["1"],{"key":"2","value":"3"},"0ca70533982928b147b5fdc377559c",{"cachedAt":1741622350514,"type":"4","metadata":"5"},"json",{"response":"6"},{"url":"7","status":200,"headers":"8"},"https://11tybundle.dev/api/global-data.json"]

